tekton/publish.yaml

```diff
@@ -0,0 +1,183 @@
+    - name: imageRegistry
+      description: TODO(#569) This is a hack to make it easy for folks to switch the registry being used by the many many image outputs
+    - name: pathToProject
+      description: The path to the folder in the go/src dir that contains the project, which is used by `ko` to name the resulting images
+  outputs:
+    resources:
+    - name: bucket
+      type: storage
+    - name: builtBaseImage
+      type: image
+    - name: builtEntrypointImage
+      type: image
+    - name: builtKubeconfigWriterImage
+      type: image
+    - name: builtCredsInitImage
+      type: image
+    - name: builtGitInitImage
+      type: image
+    - name: builtNopImage
+      type: image
+    - name: builtBashImage
+      type: image
+    - name: builtGsutilImage
+      type: image
+    - name: builtControllerImage
+      type: image
+    - name: builtWebhookImage
+      type: image
+  steps:
+
+  - name: build-push-base-images
+    image: gcr.io/kaniko-project/executor
+    command:
+    - /kaniko/executor
+    args:
+    - --dockerfile=/workspace/go/src/github.com/tektoncd/pipeline/images/Dockerfile
+    - --destination=${inputs.params.imageRegistry}/${inputs.params.pathToProject}/${outputs.resources.builtBaseImage.url}
+    - --context=/workspace/go/src/github.com/tektoncd/pipeline
+
+  - name: create-ko-yaml
+    image: busybox
+    command:
+    - /bin/sh
+    args:
+    - -ce
+    - |
+      set -e
+      set -x
+
+      cat <<EOF > /workspace/go/src/github.com/tektoncd/pipeline/.ko.yaml
+      baseImageOverrides:
+        ${inputs.params.pathToProject}/${outputs.resources.builtCredsInitImage.url}: ${inputs.params.imageRegistry}/${inputs.params.pathToProject}/build-base:latest
+        ${inputs.params.pathToProject}/${outputs.resources.builtGitInitImage.url}: ${inputs.params.imageRegistry}/${inputs.params.pathToProject}/build-base:latest
+        ${inputs.params.pathToProject}/${outputs.resources.builtBashImage.url}: busybox # image should have shell in $PATH
+        ${inputs.params.pathToProject}/${outputs.resources.builtGsutilImage.url}: google/cloud-sdk:alpine # image should have gsutil in $PATH
+      EOF
+
+      cat /workspace/go/src/github.com/tektoncd/pipeline/.ko.yaml
+
+  # TODO(#631): publish a `ko` image
+  - name: install-ko
+    image: golang
+    env:
+    - name: GOBIN
+      value: /workspace
+    command: ["go"]
+    args:
+    - "get"
+    - "github.com/google/go-containerregistry/cmd/ko"
+
+  # TODO(#631): publish a `ko` image (which has golang)
+  - name: install-go
+    image: golang
+    env:
+    command:
+    - /bin/sh
+    args:
+    - -ce
+    - |
+      set -e
+      set -x
+      # TODO(#631): this is a hack to make the go binary available in a container that has gcloud
+      mkdir -p /workspace/golang
+      cp /usr/local/go/bin/go /workspace/golang/go
+      cp -R /usr/local/go /workspace/golang/localgo
+
+  - name: ensure-release-dirs-exist
+    image: busybox
```

Comment:
when the bucket was empty, I got the error
```
error exit status 1; cmd_output CommandException: No URLs matched: gs://pipeline-test-123456/*\n","stacktrace":"main.main\n\t/Users/pivotal/go/src/github.com/tektoncd/pipeline/cmd/gsutil/main.go:94\nruntime.main\n\t/usr/local/Cellar/go/1.12/libexec/src/runtime/proc.go:200"
```

but that might be a bug in the resource itself

Reply:
oh interesting - so you'd created the bucket but nothing was in it? ill try that out too

Reply:
hm okay so it seems like if you use the `dir` flag and the bucket is empty, you get this error:

```
Error executing command "gsutil cp -r gs://christies-empty-bucket/* /workspace/bucket" ; error exit status 1; cmd_output CommandException: No URLs matched: gs://christies-empty-bucket/*
"   
  stacktrace:  "main.main
	/Users/christiewilson/Code/go/src/github.com/tektoncd/pipeline/cmd/gsutil/main.go:94
runtime.main
	/usr/local/Cellar/go/1.10.3/libexec/src/runtime/proc.go:198"   
  ts:  1553120763.622061   
```

My feeling is that this is a bug in the resource; I think we should be able to handle checking out an empty bucket that we intend to put content into.

Reply:
Created https://github.com/tektoncd/pipeline/issues/646